db.go

```diff
@@ -767,17 +779,49 @@ func (db *DB) Close() error {
 		}
 	}
 
+	if err := db.closeInternal(); err != nil {
+		return err
+	}
+
 	if len(db.sinks) != 0 {
 		// If we've successfully persisted all the table blocks we can remove
-		// the wal and snapshots.
-		if err := os.RemoveAll(db.snapshotsDir()); err != nil {
-			return err
+		// the wal and snapshots. Move the two directories to a trash dir since
+		// it is an O(1) operation. The trash dir is cleaned up on startup in
+		// case we cannot delete the WAL/snapshots in time.
+		trashDir := db.trashDir()
+
+		if moveErr := func() error {
+			_ = os.Mkdir(trashDir, os.FileMode(0o755))
+			// Create a temporary directory in the trash dir to avoid clashing
+			// with other wal/snapshot dirs that might not have been removed
+			// previously.
+			tmpPath, err := os.MkdirTemp(trashDir, "")
+			if err != nil {
+				return err
+			}
+			if err := os.Rename(db.snapshotsDir(), filepath.Join(tmpPath, snapshotsPath)); err != nil && !os.IsNotExist(err) {
+				return err
+			}
+			if err := os.Rename(db.walDir(), filepath.Join(tmpPath, walPath)); err != nil && !os.IsNotExist(err) {
+				return err
+			}
+			return nil
+		}(); moveErr != nil {
+			// If we failed to move the wal/snapshots to the trash dir, fall
+			// back to attempting to remove them with RemoveAll.
+			if err := os.RemoveAll(db.snapshotsDir()); err != nil {
+				return fmt.Errorf("%v: %v", moveErr, err)
+			}
+			if err := os.RemoveAll(db.walDir()); err != nil {
+				return fmt.Errorf("%v: %v", moveErr, err)
+			}
+			return moveErr
 		}
-		if err := os.RemoveAll(db.walDir()); err != nil {
+		if err := os.RemoveAll(trashDir); err != nil {
```

Comment:
Just for my understanding: This is what might take too long and we don't finish in time to leave data dangling?